includes/utils/class-amp-dom-document.php

```diff
@@ -0,0 +1,313 @@
+			]
+		);
+		$head->insertBefore( $charset, $head->firstChild );
+
+		return str_replace( '<meta http-equiv="content-type" content="text/html; charset=' . self::AMP_ENCODING . '">', '', parent::saveHTML( $node ) );
+	}
+
+	/**
+	 * Maybe add the <head> and/or <body> tag(s).
+	 *
+	 * @param string $content Content to add the head or body tags to.
+	 * @return string Adapted content.
+	 */
+	private function maybe_add_head_or_body( $content ) {
+		$substring = substr( $content, 0, 5000 );
+		if ( false === strpos( $substring, '<body' ) ) {
+			if ( false === strpos( $substring, '<head>' ) ) {
+				// Create the required HTML structure if none exists yet.
+				$content = "<html><head></head><body>{$content}</body></html>";
+			} else {
+				// <head> seems to be present without <body>.
+				$content = preg_replace( '#</head>(.*)</html>#', '</head><body>$1</body>', $content );
+			}
+		} elseif ( false === strpos( $substring, '<head>' ) ) {
+			// Create a <head> element if none exists yet.
+			$content = str_replace( '<body', '<head></head><body', $content );
+		}
+
+		return $content;
+	}
+
+	/**
+	 * Adapt the encoding of the content.
+	 *
+	 * @param string $source Source content to adapt the encoding of.
+	 * @return string Adapted content.
+	 */
+	private function adapt_encoding( $source ) {
+		// No encoding was provided, so we need to guess.
+		if ( self::UNKNOWN_ENCODING === $this->original_encoding ) {
+			$this->original_encoding = mb_detect_encoding( $source );
+		}
+
+		// Guessing the encoding seems to have failed, so we assume UTF-8 instead.
+		if ( empty( $this->original_encoding ) ) {
+			$this->original_encoding = self::AMP_ENCODING;
+		}
+
+		$this->original_encoding = $this->sanitize_encoding( $this->original_encoding );
+
+		$target = mb_convert_encoding( $source, self::AMP_ENCODING, $this->original_encoding );
+
+		return is_string( $target ) ? $target : $source;
+	}
+
+	/**
+	 * Detect the encoding of the document.
+	 *
+	 * @param string $content Content of which to detect the encoding.
+	 * @return string|false Detected encoding of the document, or false if none.
+	 */
+	private function detect_and_strip_encoding( &$content ) {
+		$encoding = self::UNKNOWN_ENCODING;
+
+		// Check for HTML 4 http-equiv meta tags.
+		$http_equiv_tag = $this->find_tag( $content, 'meta', 'http-equiv' );
+		if ( $http_equiv_tag ) {
+			$encoding = $this->extract_value( $http_equiv_tag, 'charset' );
+		}
+
+		// Check for HTML 5 charset meta tag. This overrides the HTML 4 charset.
+		$charset_tag = $this->find_tag( $content, 'meta', 'charset' );
+		if ( $charset_tag ) {
+			$encoding = $this->extract_value( $charset_tag, 'charset' );
+		}
+
+		// Strip charset tags if they don't fit the AMP UTF-8 requirement.
+		if ( self::AMP_ENCODING !== strtolower( $encoding ) ) {
+			$http_equiv_tag && str_replace( $http_equiv_tag, '', $content );
+			$charset_tag && str_replace( $charset_tag, '', $content );
```

Comment:
These seem to be unused expressions. Shouldn't they be assigned to something?

Reply:
These are string replacements that only happen when the strings evaluate to `true`. Too cryptic?

Reply:
The string replacements aren't done in place though, as `$content` is not passed by reference. I would expect something like this:

```php
if ( $charset_tag ) {
    $content = str_replace( $charset_tag, '', $content );
}
```

Reply:
You're right, this is nonsense. I'll change it.